plugins/cc/cctac.ml

```diff
@@ -475,9 +488,10 @@ let cc_tactic depth additionnal_terms =
           convert_to_hyp_tac ida ta idb tb p
   end
 
-
 let congruence_tac depth l =
-  Tacticals.New.tclTHEN (Tacticals.New.tclREPEAT introf) (cc_tactic depth l)
+  Tacticals.New.tclTHEN
+    (Tacticals.New.tclREPEAT (Tacticals.New.tclFIRST [intro; Tacticals.New.tclTHEN hnf_in_concl intro]))
```

Comment:
I think it's a bad idea to use `hnf_in_concl` here, even if it corresponds to the inlining of `intro`. It's a very weird reduction tactic that should be phased out.  The way this is written is probably going to be inefficient as well. Since we're going for a change of semantics anyways, I'd rather recommend to do what `intro` does, i.e. match on the goal directly but try to `whd`-reduce if not a product / let-in.

Reply:
I was mostly following @herbelin's suggestion https://github.com/coq/coq/issues/14662#issuecomment-880867794.
Since `intro` (as opposed to `introf`) is fast and well behaved, I could (instead of copy-pasting `intro` code) replace in the above `hnf_in_concl` by a custom `whd_in_concl` which is basically `Reductionops.clos_whd_flags CClosure.all env sigma t`.
Trying `intro` twice in the reduction case, the performance hit (for `congruence`, which itself is far more expensive) is negligible.

Reply:
I somehow agree with @ppedrot that "`hnf_concl` is a very weird reduction tactic that should be phased out". I proposed to use it because it is what `intro` does already. A form of `whd_in_concl` which preserves let-ins is probably good.

Reply:
There is a dedicated flag for that, so `Reductionops.clos_whd_flags CClosure.allnolet` is probably the desired reduction.

Reply:
I tried `Reductionops.clos_whd_flags CClosure.allnolet` but it broke a lot of use cases such as proving `Proper (le ==> le) f.` by `congruence`. For this, weak head without let does not produce a good conclusion for a subsequent `intro`.